MbsAPI/fLmd.c

```diff
@@ -61,1186 +56,1266 @@ int clock_gettime(int clockid, struct timespec* tp)
-      }
 
-      // second, try to read more bytes
+    bufferBytes = iBytes;
+    if (bufferBytes < pLmdControl->pMbsFileHeader->iMaxWords * 2
+        && (pLmdControl->pMbsFileHeader->iMaxWords < UINT32_MAX / 2)) {
+        bufferBytes = pLmdControl->pMbsFileHeader->iMaxWords * 2;
+    }
+    fLmdPrintFileHeader(1, pLmdControl->pMbsFileHeader);
+    pLmdControl->pBuffer = (int16_t*)malloc(bufferBytes);
+    pLmdControl->iBufferWords = bufferBytes / 2;   // will be increased if necessary
 
-      iReturn = fLmdReadBuffer(pLmdControl,
-                               (char*)(pLmdControl->pBuffer+pLmdControl->iLeftWords),
-                               (pLmdControl->iBufferWords-pLmdControl->iLeftWords)*2);
+    printf("fLmdGetOpen: %s words %u\n", Filename, pLmdControl->iBufferWords);
 
-      if(iReturn <= 0) { printf("fLmdGetElement: EOF\n"); return(GETLMD__EOFILE); }
+    pLmdControl->iLeftWords = 0;   // buffer empty, read with first fLmdGetElement
+    pLmdControl->pMbsHeader = NULL;
+    return (LMD__SUCCESS);
+}
+//===============================================================
+uint32_t fLmdGetBuffer(sLmdControl* pLmdControl,
+                       sMbsHeader* pMbsHeader,
+                       uint32_t iBytes,
+                       uint32_t* iElements,
+                       uint32_t* iBytesUsed)
+{
 
-      if(pLmdControl->iSwap) { fLmdSwap4((uint32_t*)(pLmdControl->pBuffer+pLmdControl->iLeftWords),iReturn/4); }
+    sMbsHeader* pm;
+    uint32_t elem = 0, leftBytes = 0, used, elem_sz;   //, *ps, *pd, i,ii, size=0;
+    int32_t iReturn;
 
-      pLmdControl->iBytes += iReturn;
-      pLmdControl->pMbsHeader=(sMbsHeader*)pLmdControl->pBuffer;
-      pLmdControl->iLeftWords += iReturn/2;
+    if (iBytes < pLmdControl->pMbsFileHeader->iMaxWords) {
+        printf("fLmdGetBuffer: %s buffer size %d too small for %d bytes\n",
+               pLmdControl->cFile,
+               iBytes,
+               pLmdControl->pMbsFileHeader->iMaxWords);
+        return (LMD__FAILURE);
+    }
+    if (pMbsHeader == NULL) {
+        printf("fLmdGetBuffer: Need buffer to read\n");
+        return (LMD__FAILURE);
+    }
+    *iBytesUsed = 0;
+    *iElements = 0;
+    if (pLmdControl->iElements == pLmdControl->pMbsFileHeader->iElements) {
+        return (GETLMD__EOFILE);
     }
 
-    // check if read buffer enough for event
+    // Offset table
+    if (pLmdControl->iOffsetEntries) {   // use offsets to read elements fitting in buffer
+        fLmdOffsetElements(pLmdControl, iBytes, &elem, &used);
+        // printf("Read %d bytes of %d, elements %d\n",used,iBytes,elem);
+        iReturn = fLmdReadBuffer(pLmdControl, (char*)pMbsHeader, used);
+        if (iReturn <= 0) {
+            printf("fLmdGetBuffer: EOF: %s\n", pLmdControl->cFile);
+            return (GETLMD__EOFILE);
+        }
+        if (iReturn != used) {
+            printf("fLmdGetBuffer: LMD read error: unexpected EOF: %s %u %u\n", pLmdControl->cFile, iReturn, used);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [%u in format string (no. 2) requires 'unsigned int' but the argument type is 'signed int'.](https://app.codacy.com/manual/FairRootGroup/FairRoot/pullRequest?prid=5323596)